File: alura/php-design-patterns/IHIT.php
<?php
  class IHIT extends TemplateDeImpostoCondicional {
    protected function deveUsarMaximaTaxacao(Orcamento $orcamento) {
      $noOrcamento = Array();

      foreach($orcamento->getItens() as $item) {
        if(in_array($item->getNome(),$noOrcamento)) {
          return true;
        } else {
          $noOrcamento[] = $item->getNome();
        }
      }
      return false;
    }

    protected function maximaTaxacao(Orcamento $orcamento) {
      return $orcamento->getValor() * 0.13 + 100;
    }

    protected function minimaTaxacao(Orcamento $orcamento) {
      return $orcamento->getValor() * (0.01 * count($orcamento->getItens()));
    }
  }
?>

File: alura/java/design-patterns/src/IHIT.java
class IHIT extends TemplateDeImpostoCondicional {

  public boolean deveUsarMaximaTaxacao(Orcamento orcamento) {
    List<String> noOrcamento = new ArrayList<String>();

    for(Item item : orcamento.getItens()) {
      if(noOrcamento.contains(item.getNome())) {
        return true;
      } else {
        noOrcamento.add(item.getNome());
      }
    }

    return false;
  }

  public abstract double maximaTaxacao(Orcamento orcamento) {
    return orcamento.getValor() * 0.13 + 100;
  }

  public abstract double minimaTaxacao(Orcamento orcamento) {
    return orcamento.getValor() * (0.01 * orcamento.getItens().size());
  }
}
